gdal/ogr/ogrsf_frmts/shape/shpopen.c

```diff
@@ -589,20 +598,110 @@ SHPOpenLL( const char * pszLayer, const char * pszAccess, SAHooks *psHooks )
         return NULL;
     }
 
+/* -------------------------------------------------------------------- */
+/*  Read the file size from the SHP file.                               */
+/* -------------------------------------------------------------------- */
+    pabyBuf = (uchar *) malloc(100);
+    psSHP->sHooks.FRead( pabyBuf, 100, 1, psSHP->fpSHP );
+    
+    psSHP->nFileSize = ((unsigned int)pabyBuf[24] * 256 * 256 * 256
+                        + (unsigned int)pabyBuf[25] * 256 * 256
+                        + (unsigned int)pabyBuf[26] * 256
+                        + (unsigned int)pabyBuf[27]);
+    if( psSHP->nFileSize < 0xFFFFFFFFU / 2 )
+        psSHP->nFileSize *= 2;
+    else
+        psSHP->nFileSize = 0xFFFFFFFEU;
+    
+/* -------------------------------------------------------------------- */
+/*	Create the .shx file for existing .shp, if SHP_RESTORE_SHX parameter*/
+/*	was passed.                                                         */
+/* -------------------------------------------------------------------- */
+    if( bSHXRestoring && psSHP->fpSHX == NULL )
+    {
+        const char pszSHXAccess[] = "w+b";
+        snprintf( pszFullname, nFullnameLen, "%s.shx", pszBasename );
+        psSHP->fpSHX = psHooks->FOpen( pszFullname, pszSHXAccess );
+        
+        psHooks->FSeek( psSHP->fpSHP, 100, 0 );
+        char *pabySHXContent = (char *) malloc ( psSHP->nFileSize );
```

Comment:
This looks to be a really excessive allocation ! And the returned pointer should be checked against NULL.